### PR TITLE
docs: ADR to use React Query instead of Redux, update general JS/TS guidance

### PR DESCRIPTION
**Pull request by bradenmacdonald**
* Changed guidance on React state/data loading to recommend React Query instead of Redux
* Updated JavaScript/TypeScript guidance

Previous discussions etc:
* https://openedx.atlassian.net/wiki/spaces/BPL/pages/3852632086/Typescript+and+React-Query+for+api+usage.
* https://openedx.atlassian.net/wiki/spaces/AC/pages/3791290378/tanstack+react-query+An+overview
* My conference [lightning talk on React Query](https://docs.google.com/presentation/d/19AIPaRDJkOMLnCn-7BapHNLpS1Hz9Pti9Rz-apI5ZVY/edit)

Preview this at https://open-edx-proposals--616.org.readthedocs.build/en/616/best-practices/oep-0067-bp-tools-and-technology.html

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@sarina`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Comment by sarina** [comment]
It would be great to have something in this tree, I think: https://docs.openedx.org/en/latest/developers/references/developer_guide/conventions/index.html that points to this OEP. Or maybe there's a better place within https://docs.openedx.org/en/latest/developers/references/developer_guide/index.html? What do you think?

**Comment by bradenmacdonald** [comment]
Sure. Maybe in both places?

**Comment by itsjeyd** [comment]
@sarina It looks like you already started reviewing this PR, so I went ahead and assigned the review to you. Let me know if that's not what you had in mind.

**Comment by sarina** [comment]
@bradenmacdonald yeah in both places sounds good.

@itsjeyd - I am the maintainer of this repo so I commented a bit on structure, but I'm not a frontend expert so Braden should seek to find a reviewer from that group.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (brian-smith-tcril) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:134]: Would it be worth linking to https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0067/decisions/frontend/0008-typescript.html here?

**Review by brian-smith-tcril — APPROVED** [review]
This looks great! I'm 100% on board with recommending React Query, and the reasoning provided here is clear and persuasive.

I left one tiny non-blocking comment, but LGTM!

**Comment by bradenmacdonald** [comment]
> yeah in both places sounds good.

@sarina Since those are in a different repo, I'll open a PR to link to this from both places once this PR has merged.

I think this should be good to merge soon, but [I've posted on Slack one last time to see if anyone else wants to comment/review](https://openedx.slack.com/archives/C04BM6YC7A6/p1723227152789459?thread_ts=1721943178.602739&cid=C04BM6YC7A6).

**Review by arbrandes — APPROVED** [review]
100%.

**Review by adamstankiewicz — APPROVED** [review]
Glad to to see React Query make its way into the OEPs :)

**Comment by bradenmacdonald** [comment]
Thanks everyone!

**Comment by openedx-webhooks** [comment]
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/open-edx-proposals&entry.867055334=https://github.com/openedx/open-edx-proposals/pull/616&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2024-07-25+21:29&entry.1917517419=2024-08-15+16:47&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
